File: 09-intervals/car-pooling/solution.js
function carPooling(trips, capacity) {
    const events = [];

    for (const [passengers, start, end] of trips) {
        events.push([start, passengers]);
        events.push([end, -passengers]);
    }

    // At the same location, drop-offs should happen before pick-ups.
    events.sort((a, b) => a[0] - b[0] || a[1] - b[1]);

    let currentLoad = 0;
    for (const [, change] of events) {
        currentLoad += change;
        if (currentLoad > capacity) {
            return false;
        }
    }

    return true;
}

const trips = [[2, 1, 5], [3, 3, 7]];

console.log(`capacity = 4 -> ${carPooling(trips, 4)}`);
console.log(`capacity = 5 -> ${carPooling(trips, 5)}`);

File: 09-intervals/car-pooling/solution.py
from typing import List, Tuple


def car_pooling(trips: List[List[int]], capacity: int) -> bool:
    events: List[Tuple[int, int]] = []

    for passengers, start, end in trips:
        events.append((start, passengers))
        events.append((end, -passengers))

    # At the same location, drop-offs should happen before pick-ups.
    events.sort()

    current_load = 0
    for _, change in events:
        current_load += change
        if current_load > capacity:
            return False

    return True


if __name__ == "__main__":
    trips = [[2, 1, 5], [3, 3, 7]]

    print(f"capacity = 4 -> {car_pooling(trips, 4)}")
    print(f"capacity = 5 -> {car_pooling(trips, 5)}")
